### PR TITLE
AnvilDamagedEvent

### PR DESCRIPTION
**Pull request by BillyGalbreath**
I'm bored.

### CODE REVIEW

**Comment by scrayos** [comment]
Those events are amazing! I was always searching for something like that. I'm currently work-arounding it with Packet-Listeners, and I would love to replace them with bukkit-events.

File: Spigot-Server-Patches/0267-More-Anvil-Events.patch [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+diff --git a/src/main/java/net/minecraft/server/ContainerAnvil.java b/src/main/java/net/minecraft/server/ContainerAnvil.java
+index 16ec67569..7bcfd64e2 100644
+--- a/src/main/java/net/minecraft/server/ContainerAnvil.java
++++ b/src/main/java/net/minecraft/server/ContainerAnvil.java
+@@ -74,13 +74,22 @@ public class ContainerAnvil extends Container {
+                     int i = ((Integer) iblockdata.get(BlockAnvil.DAMAGE)).intValue();
+ 
+                     ++i;
+-                    if (i > 2) {
```
Comment (aikar) [inline Spigot-Server-Patches/0267-More-Anvil-Events.patch]: Can we avoid the indentation change? Cant see context below, but could we return early if the damage event is cancelled?

File: Spigot-Server-Patches/0267-More-Anvil-Events.patch [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+-                        world.setAir(blockposition);
+-                        world.triggerEffect(1029, blockposition, 0);
+-                    } else {
+-                        world.setTypeAndData(blockposition, iblockdata.set(BlockAnvil.DAMAGE, Integer.valueOf(i)), 2);
```
Comment (aikar) [inline Spigot-Server-Patches/0267-More-Anvil-Events.patch]: Again here, can we return early instead if cancelled?

File: Spigot-API-Patches/0085-More-Anvil-Events.patch [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
++        this.damage = damage;
++    }
++
++    @Override
++    public AnvilInventory getInventory() {
++        return (AnvilInventory) super.getInventory();
++    }
++
++    /**
++     * Gets the anvil damage value
```
Comment (aikar) [inline Spigot-API-Patches/0085-More-Anvil-Events.patch]: Can you make it more clear that this is the NEW damage value

File: Spigot-API-Patches/0085-More-Anvil-Events.patch [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
++     * <p>
++     * 0 = Not damaged
++     * 1 = Slightly damaged
++     * 2 = Very damaged
++     * 3+ = Anvil is breaking
```
Comment (aikar) [inline Spigot-API-Patches/0085-More-Anvil-Events.patch]: 3 being the breaking point is an implementation detail.

I think we should consolidate the 2 events into this one, and add a isBreaking method that returns true if damage >= 3, and false if < 3.

Then we can add a setBreaking(boolean) that simply sets damage level to 3 if true, else leave as is if < 3, and set to 2 if was 3 before.

File: Spigot-API-Patches/0085-More-Anvil-Events.patch [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
++import org.bukkit.event.Cancellable;
++import org.bukkit.event.HandlerList;
++import org.bukkit.event.inventory.InventoryEvent;
++import org.bukkit.inventory.AnvilInventory;
++import org.bukkit.inventory.InventoryView;
++
++/**
++ * Called when an anvil is broken from being used
++ */
++public class AnvilBreakEvent extends InventoryEvent implements Cancellable {
```
Comment (aikar) [inline Spigot-API-Patches/0085-More-Anvil-Events.patch]: See comments on the other event, i think we can remove this event.

**Review by aikar — CHANGES_REQUESTED** [review]

File: Spigot-Server-Patches/0267-More-Anvil-Events.patch (hunk 2 above)
Comment (aikar) [reply Spigot-Server-Patches/0267-More-Anvil-Events.patch]: Though with a merged event, you wont even need to change anything about this.

**Comment by BillyGalbreath** [comment]
Changes made as requested

**Comment by Black-Hole** [comment]
Since 1.13 is nearly complete this PR should wait, since in 1.13 anvil will be split into three different blocks.

**Comment by BillyGalbreath** [comment]
Since 1.13 is NOT nearly here, lets NOT wait. ;)

**Comment by zachbr** [comment]
I can understand why some may want this held back until 1.13.

This event is pretty well based around magic data values.
My understanding is that these values will almost certainly be changing come 1.13. Even if not moving to some form of state, changing nonetheless. If 1.13 migrates these to a more concrete state, whether that be blockstates or just new values, we could probably abstract those magic values into something with actual meaning.

Do we want to introduce this event for ... a month, month and a half, w/e only to have all the data it uses change and force plugins to update then anyway?

**Comment by BillyGalbreath** [comment]
I dont mind if it gets delayed until 1.13, I dont even use this. I was working on something else in the code here (an anvil repair event that didnt work out) and ended up adding this out of boredom/defeat. Figured since the 1.12->1.13 update is going to be another one of those 1.8->1.9 updates that make a huge chunk of the community not update it would still get used for quite some time in 1.12 (I'm not in that crowd, I'm ready for 1.13 already) ^_^

**Comment by scrayos** [comment]
Yeah, I also think this can wait until 1.13, but I really hope once 1.13 is out, this PR will still be remembered and it will make it into PaperSpigot. There is waaay too little control over anvils now, and I am absolutely looking forward to having this within PaperSpigot.

**Comment by BillyGalbreath** [comment]
6 months later... :P

**Comment by MiniDigger** [comment]
1.13 will come out eventually :D

**Pull request closed without merging**